feat(lib-dynamodb-data-mapper): extend data mapper functionality with DataMarshaller

apopovicius opened the pull request:
### Description
This contains following changes:
- datamarshaller implementation and helpers that provide marshalling from  object prototype schema
- reuse the command input/output and commands from @aws/lib-dynamodb
- extend and change unit test already in place
- update package.json findings from previous review
- created a public user interface for the data mapper
- created the first version of documentation in the readme file
- **additional**: decorators support, tableNameResolver utility 

### Testing
Minimal unit tested added on this PR. Additional will follow as soon as more code is brought here.

### Additional context
This is just the project init and more code and PR will follow up. 

### Checklist
- [x] If you wrote E2E tests, are they resilient to concurrent I/O?
- [x] If adding new public functions, did you add the `@public` tag and enable doc generation on the package?

apopovicius committed:
feat(lib): DataMarshaller implementation

extend and fix existing unit test

implement marshaller mw to be used in datamapper before each command

port expressions to lib data mapper to comply to js sdk v3

remove unnecessary commands as we are using the lib-dynamodb commands

apopovicius committed:
fix(lib): fix mathematical expression unit test

apopovicius committed:
chor(lib): create the initial version of the DataMapper readme file

apopovicius committed:
chore(lib): rename define model sections to proper meaning

apopovicius committed:
chore(lib): improve decorator example

raduendv commented on lib/lib-dynamodb-data-mapper/src/DataMapper.ts:
```diff
@@ -12,13 +20,14 @@ export class DataMapper {
 
   constructor(config: DataMapperConfig) {
     this.client = config.docClient;
+    registerSchemaMiddleware(this.client);
```
- must check for duplication of middleware registration
- how does this affect calls from outside the data mapper to the client since the middlewares are registered on the client?

apopovicius committed:
feat: rework on DataMapper wrapper

remove the ported functionality thats not used in the existing code base

reshape the interface and update the documentation

apopovicius committed:
feat: remove empty file

apopovicius committed:
chore: remove my notes file from project

apopovicius committed:
feat: fix failing unit test and adjust

apopovicius committed:
chore: follow the naming convention from JS SDK V3

apopovicius committed:
chore: fix tsdoc/syntax errors

apopovicius committed:
feat: document the DataMapper wrapper to allign with JS SDK conventions

remove interface file as its not allign with JS SDK V3

apopovicius committed:
feat: fix consistency in DataMapper: both put() and get() now rely on metadata located on the prototype.

adjust datamapper unit test to new changes

fix the end2end test

kuhe commented on lib/lib-dynamodb-data-mapper/README.md:
```diff
@@ -1,132 +1,471 @@
+   * (e.g., timeouts, retries, abort signals). See {@link HttpHandlerOptions}.
+   * @returns A hydrated instance of the model.
+   */
+  get<T extends object>(key: Partial<T>, modelCtor: new () => T, options?: HttpHandlerOptions): Promise<T | undefined>;
```
`new () => T` requires a class have a no-args constructor. This isn't very flexible. Are you sure this is how users want to interact with the library?

It should accept any factory function of the signature 
```ts
<D, T>(data: D) => T;
```

kuhe commented on lib/lib-dynamodb-data-mapper/README.md:
```diff
@@ -1,132 +1,471 @@
+    modelCtor: new () => T,
+    criteria?: Omit<Partial<DeleteItemCommandInput>, "TableName" | "Key">,
+    options?: HttpHandlerOptions
+  ): Promise<T>;
```
why return the object itself? why not a boolean for if the deletion succeeded?

the comment above says "returns deleted item if it existed". What if it didn't?

kuhe commented on lib/lib-dynamodb-data-mapper/README.md:
```diff
@@ -1,132 +1,471 @@
+   * @returns An async iterable of hydrated model instances.
+   */
+  query<T extends object>(
+    modelCtor: new () => T,
```
isn't an instance of datamapper associated with a single table and single object schema?

`modelCtor`, or the generalized factory function should be applied in the constructor rather than for every method.

kuhe commented on lib/lib-dynamodb-data-mapper/CONTRIBUTING.md:
```diff
@@ -0,0 +1,181 @@
+# Contributing to @aws-sdk/lib-dynamodb-data-mapper
+
+Welcome! We're excited that you're interested in contributing to the DataMapper for AWS SDK JavaScript v3. This project aims to offer a clean, modular, and schema-aware abstraction over DynamoDB operations. Please follow the guidelines below to help us maintain quality and consistency.
```
we use the word "Schema" to refer to an existing concept in the internals of the SDK, please rename all usage of the word schema to "DynamoDBItemSchema" or something more specific.

apopovicius replied:
I was thinking to use ItemSchema and ItemType. What do you think? Do we need to be refer to DataMapper or DynamodDB?

kuhe replied:
ItemSchema sounds good to me

kuhe commented on lib/lib-dynamodb-data-mapper/README.md:
```diff
@@ -1,132 +1,471 @@
-- middleware/ Not applicable yet (but possible in future)
+  @attribute()
+  city!: string;
+}
```
As I mentioned above, this system assumes all users work with mutable noArgsConstructor data object classes. 

We should support a factory function definition:

```ts
function factory<D, T>(data: D): T;

new DataMapper<D, T>({
  factory
});
```

There needs to be two data-carrier types defined, the base object type that is retrieved by the SDK, and the mapped class type that the user receives. You can to some extent infer `D` from the constructor signature of `T`, but it may not even be a class, and even if it is a class it may have a constructor pattern that doesn't involve the `new` operator. 

Imagine that the user desired type is immutable and only has a static constructor:
```ts
new DataMapper<DataType, MyObject>({
  factory(data: DataType) {
    return MyObject.from(data);
  }
});
```

kuhe replied:
two functions need to be configurable, `D -> T` for untyped data to typed data, and `T -> D` for typed data to untyped data.

apopovicius committed:
feat: remove compiler option related to decorators from tsconfig.cjs .es

apopovicius committed:
chore: rename schema and schema types for consistency

apopovicius committed:
fix: update versionion to fix husky precommit error

apopovicius committed:
feat: unify schema and marshaller structure, introduce TableMapper API

- Moved schema files (itemSchema, itemType, itemKeySchema) into schema/item/
- Renamed schema types to ItemSchema, ItemType to avoid conflicts with SDK internals
- Relocated marshallKey to marshaller/, grouped all marshalling logic under one module
- Removed hydration logic from marshaller layer (unmarshallObject now returns plain objects)
- Updated unmarshallObject to recursively flatten embedded documents
- Added support for factory-based model construction in DataMapper.get<D, T>()
- Introduced TableMapper.from(...) for model-bound DX-friendly mapping
- Reorganized test structure to reflect new folder layout
- Preserved backward-compatible overloads for legacy get(modelCtor) usage

BREAKING CHANGE:
- DataMapper.get() now requires a factory and modelCtor;
 legacy overload still supported for transition
- unmarshallObject() no longer hydrates embedded classes automatically

apopovicius committed:
feat: add unmarshallMap and isNullValue helper for safer and DRYer decoding

apopovicius committed:
feat: refactor marshaller layout and improve marshall/unmarshall symmetry

- Split core into marshall and unmarshall folders
- Extracted marshallDocument and unmarshallDocument
- Improved symmetry between marshallItem and unmarshallItem
- Moved AttributeValueMap to internal/dynamodb-types.ts
- Updated DataMarshaller with cleaner API and docs

apopovicius committed:
fix(unmarshallItem): apply defaultProvider fallback for missing fields

- If a field is missing in the DynamoDB item and has a defaultProvider,
  the default value is now assigned during unmarshalling
- Fixes failing test for Document schema with defaultProvider
- Aligns unmarshallItem behavior with marshallItem symmetry

apopovicius committed:
feat: for some reason the previous commit didnt took marshall subfolder

apopovicius committed:
fix: marshall value typesafety check for string, binary, any

created a test to validate that ts decorators works in the current setup

adapt config files to work with decorators

apopovicius committed:
feat: support dual get/put interfaces and introduce public command input types

- Add GetCommandInput and PutCommandInput for structured, object-style usage
- Support both classic and structured interfaces for get() and put()
- Refactor argument resolution logic into helper
- Align API with implementation progress and SDK v3 idioms
- Update tests for both constructor and factory hydration flows

apopovicius commented:
As there are multiple discussion around the public interface, I will move the PR in draft since the code will be aligned with the multiple use-cases discussed during review meeting.

apopovicius committed:
feat(datamapper): implement flexible schema-aware DataMapper with unified test coverage

- Refactored DataMapper to support multiple usage scenarios:
  - Plain object (T = D)
  - Class with manual transformers
  - Class with static from/toDocument methods
  - Protocol-based schema metadata
  - (Stubbed) decorator-based schema

- Added static factory method `from()` to enforce DX and encapsulate logic

- Implemented unit tests for:
  - DataMapper.from() wiring validation
  - put() and get() with mocked client.send()
  - All use cases run per-method
  - delete/query/scan throw not implemented

- Maintained full compatibility with aws-sdk v3 clients and util-dynamodb fallback

BREAKING CHANGE: direct constructor usage is now private; use DataMapper.from() instead

apopovicius committed:
chore: update readme with latest design changes

apopovicius committed:
feat(attribute): add attribute decorator and tests, skip tests requiring TS metadata flags

apopovicius committed:
fix: datamapper schema and tablename are optional

apopovicius committed:
feat(decorators): bring all v2 decorators in the new v3 datamapper

apopovicius committed:
feat(datamapper): add tableNameResolver utility

apopovicius committed:
feat: refactor tableNameResolver to make use of getTable from protocol symbol

remove unnecesary commands

bring tableNameResolver functionality in DataMapper main component

apopovicius committed:
fix: remove unnecesary exports from main index file

andre-araujo commented on lib/lib-dynamodb-data-mapper/src/marshaller/DataMarshaller.ts:
```diff
@@ -0,0 +1,73 @@
+ * This class provides static methods to transform JavaScript objects
+ * to and from DynamoDB AttributeValue format using schema definitions.
+ */
+export class DataMarshaller {
```
Why do we need this abstraction, seems like we already have all methods under marshal/

apopovicius replied:
We discuss during our internal review. I guess its fine for now and we can change it later

andre-araujo commented on lib/lib-dynamodb-data-mapper/src/marshaller/unmarshall/unmarshallPlainObject.ts:
```diff
@@ -0,0 +1,32 @@
+ * @param schema - The schema used to guide unmarshalling.
+ * @returns A plain object with unmarshalled values.
+ */
+export function unmarshallPlainObject(item: Record<string, AttributeValue>, schema: ItemSchema): Record<string, any> {
```
Why any instead of a generic?

andre-araujo commented on lib/lib-dynamodb-data-mapper/src/marshaller/unmarshall/unmarshallPlainObject.ts:
```diff
@@ -0,0 +1,32 @@
+    if (attr === undefined) continue;
+
+    if (definition.type === "Document") {
+      const docValue = (attr as any).M;
```
Not sure why we need casting here, if we already know type is document. If really needed I would prefer to have a type that represents the document value even if generic

andre-araujo commented on lib/lib-dynamodb-data-mapper/src/DataMapper.ts:
```diff
@@ -1,24 +1,231 @@
+   * Constructor is private to enforce usage of factory methods like `forModel()`.
+   */
+  private constructor(config: DataMapperConfig, binding: TableBindingConfig<D, T>) {
+    const pass = (x: any) => x;
```
Why dont we just not assign fromDocument and toDocument, then ignore instead of creating this noop function?

apopovicius replied:
We assign a pass function to avoid conditional checks every time fromDocument is called. This makes the downstream logic simpler and always safe.

andre-araujo commented on lib/lib-dynamodb-data-mapper/src/DataMapper.ts:
```diff
@@ -1,24 +1,231 @@
+      try {
+        this.schema = getSchema(binding.documentClass);
+      } catch {
+        this.schema = undefined; // fall back to schema-less mode
```
Are all errors thrown expected? I believe this generic catch may obfuscate errors in case we have an unexpected error from getSchema.

apopovicius replied:
we will address errors later in feature PRs that will also cover the user interface

andre-araujo commented on lib/lib-dynamodb-data-mapper/src/schema/tableNameResolver.ts:
```diff
@@ -0,0 +1,75 @@
+   * @param modelClass - The constructor function of the model class
+   * @returns A string representing the resolved DynamoDB table name
+   */
+  resolve(modelClass: Function): string;
```
Resolve seems like a very misleading name for a method. The usage of "resolve" from the instance is very easy to be mistaken with a promise at a quick glance.

apopovicius committed:
chore: unify schema resolution and improve decorator/runtime integration

Summary:
This commit introduces a unified and extensible mechanism for resolving table name and schema
metadata from class models, while improving TypeScript decorator behavior and test reliability.

Key Changes:

Refactored DataMapper constructor
Uses resolveModelMetadata() for consistent schema and table resolution
Supports protocol-based symbols and user-supplied overrides

Introduced resolveModelMetadata() utility
Validates presence of schema and table name metadata

Enhanced attribute() decorator
Supports Reflect metadata type inference via emitDecoratorMetadata
Handles List, Set, Map, Tuple, Document with fallback to 'Any'

Improved ConfigurableTableNameOverride
Adds withConfig(), override(), withSupplier()
Ensures override precedence over prefix/env/symbol

Fixed tests
Adjusted all prototype access to use .prototype[DynamoDbSchema]
Updated mock client behavior to reflect AWS SDK v3 (send() usage)

Fully compatible with both decorator-based and protocol-based metadata

No breaking changes.

apopovicius committed:
feat: tighten type safety in marshallers and decorators

- Removed remaining 'any' usage in marshall/unmarshall logic using generics
- Used 'typeof DynamoDbSchema' for safe symbol indexing in attribute()
- Improved typings in unmarshallItem, unmarshallMap, and unmarshallDocument
- Simplified schema container access with constrained object types

apopovicius committed:
feat: feat(schema): tighten DocumentType with generic constructor constraint

Introduce a second generic parameter  to DocumentType to ensure
valueConstructor matches the declared object type . This improves type
safety for embedded document instantiation while preserving backwards
compatibility with default .

apopovicius committed:
chore: isolate marshalling types

schema decorators extended with some functional testing

apopovicius committed:
chore: try to prevent leaking internal exports

andre-araujo commented on lib/lib-dynamodb-data-mapper/src/DataMapper.ts:
```diff
@@ -1,24 +1,236 @@
+   * Constructor is private to enforce usage of factory methods
+   */
+  private constructor(config: DataMapperConfig, binding: TableBindingConfig<D, T>) {
+    const pass = (_: any) => _;
```
We dont need this pass function, we can change this as a follow-up PR

andre-araujo commented on lib/lib-dynamodb-data-mapper/src/schema/decorators/attribute.spec.ts:
```diff
@@ -0,0 +1,394 @@
+  it("should create a document schema compatible with the DynamoDbSchema protocol", () => {
+    const decorator = attribute();
+    const target = Object.create(null);
+    console.log("decorator is", decorator);
```
Why console.log on testing?
